File: jarvis-client/app/src/main/java/com/jarvis/client/net/Focus.kt
package com.jarvis.client.net

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull

/**
 * Focus sessions (the owner's decision of 2026-09-25; docs/JARVIS-API.md
 * section 31; backend `jarvis_focus.py`, `focus.patch`).
 *
 * A timer plus Quiet. On the PC, Jarvis watches which app or site is in front
 * and names a drift out loud - on the PC only. This phone starts and stops a
 * session and shows the countdown, the counts and the report card. It never
 * sees what was in front: the PC sends only booleans and counts, and the
 * spoken line is refused to anything but the PC itself (ARCHITECTURE section
 * 8, "One-sided on purpose").
 *
 * Start, Resume and +10 minutes are held on a stale link (rule 4: they make
 * Jarvis watch, or watch longer); Pause and Stop are let through - they only
 * make it do less.
 *
 * Pure Kotlin, no Android types, so `FocusTest` runs it on a plain JVM
 * against `contract/focus-cases.json` - the PC's real answers.
 */
object Focus {
    const val PATH = "/api/focus"
    const val START_PATH = "/api/focus/start"
    const val ACT_PATH = "/api/focus/act"

    const val TITLE = "Focus session"
    const val DETAIL =
        "A timer plus Quiet. On the PC, Jarvis watches which app or site is in front " +
            "and says so when you drift - out loud, on the PC only. It keeps counts, never " +
            "what it saw, and nothing leaves the PC. Off unless you start it."
    const val PHONE_NOTE =
        "Watching what is in front happens on the PC only; this phone shows the " +
            "countdown and the counts."
    const val MISSING =
        "Your PC's Jarvis does not have focus sessions yet - run apply-patches.ps1 on the PC."

    const val DEFAULT_MINUTES = 25
    const val MIN_MINUTES = 1
    const val MAX_MINUTES = 240
    const val EXTEND_MINUTES = 10

    const val MINUTES_LABEL = "Minutes"
    const val ON_LABEL = "On what (optional)"

    /** The "On:" line while "Hide memory lists and chat history" hides what it is on. */
    const val INTENT_HIDDEN = "On: hidden until you confirm it is you."
    const val START = "Start"
    const val PAUSE = "Pause"
    const val RESUME = "Resume"
    const val EXTEND = "+10 minutes"
    const val STOP = "Stop"
    const val BAD_MINUTES = "A focus session is 1 to 240 minutes long."
    const val LAST_TITLE = "Last session"

    /** What the phone may ask for. "lock" is the desktop's alone: it is about the PC's screen. */
    val PHONE_ACTIONS = setOf("pause", "resume", "stop", "extend")

    /** Held on a stale link: they make Jarvis watch, or watch for longer. */
    val HELD_WHEN_STALE = setOf("resume", "extend")

    data class Report(
        val title: String,
        val lines: List<String>,
        val clean: Boolean,
        val streak: Int,
        val completed: Boolean,
    )

    data class View(
        val on: Boolean,
        val paused: Boolean,
        val state: String,
        val minutes: Int,
        val leftS: Long,
        val intent: String,
        val deferred: Boolean,
        val onTarget: Boolean?,
        val drifting: Boolean,
        val excused: Boolean,
        val drifts: Int,
        val line: String,
        val note: String,
        val report: Report?,
        val streak: Int,
    )

    /** `GET /api/focus`, read - or null when it is not one (an older PC). */
    fun parse(body: JsonObject): View? {
        if (body.flag("available") == false) return null
        val on = body.flag("on") ?: return null
        val rep = (body["report"] as? JsonObject)?.let { r ->
            Report(
                title = r.text("title") ?: "",
                lines = (r["lines"] as? JsonArray)?.mapNotNull { (it as? JsonPrimitive)?.contentOrNull }
                    ?: emptyList(),
                clean = r.flag("clean") == true,
                streak = r.num("streak")?.toInt() ?: 0,
                completed = r.flag("completed") == true,
            )
        }
        return View(
            on = on,
            paused = body.flag("paused") == true,
            state = body.text("state") ?: if (on) "locked" else "off",
            minutes = body.num("minutes")?.toInt() ?: 0,
            leftS = body.num("left_s")?.toLong()?.coerceAtLeast(0) ?: 0,
            intent = body.text("intent") ?: "",
            deferred = body.flag("deferred") == true,
            onTarget = body.flag("on_target"),
            drifting = body.flag("drifting") == true,
            excused = body.flag("excused") == true,
            drifts = body.num("drifts")?.toInt() ?: 0,
            line = body.text("line") ?: "",
            note = body.text("note") ?: "",
            report = rep,
            streak = body.num("streak")?.toInt() ?: 0,
        )
    }

    /** A read that failed because this PC has no focus sessions. */
    fun missing(error: ApiError): Boolean =
        error == ApiError.NotFound || error == ApiError.NotAvailable ||
            (error is ApiError.Server && error.code == 501)

    /** Seconds left now, counted down from what the PC said [sinceMs] ago. */
    fun leftNow(v: View, sinceMs: Long): Long =
        if (!v.on || v.paused) v.leftS else (v.leftS - sinceMs / 1000).coerceAtLeast(0)

    /** "24:05", or "1:02:05" past an hour. */
    fun clock(seconds: Long): String {
        val s = seconds.coerceAtLeast(0)
        val h = s / 3600
        val m = (s % 3600) / 60
        val sec = s % 60
        return if (h > 0) "%d:%02d:%02d".format(h, m, sec) else "%d:%02d".format(m, sec)
    }

    /** The minutes typed, or null when they are not 1 to 240. */
    fun minutesOf(text: String): Int? =
        text.trim().toIntOrNull()?.takeIf { it in MIN_MINUTES..MAX_MINUTES }

    /** `POST /api/focus/start`'s body, or null for minutes out of range. */
    fun startBody(minutes: Int, on: String): String? {
        if (minutes !in MIN_MINUTES..MAX_MINUTES) return null
        val words = on.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ").take(60)
        return "{\"minutes\":$minutes,\"on\":" + JarvisJson.encodeToString(
            kotlinx.serialization.serializer<String>(), words) + "}"
    }

    /** `POST /api/focus/act`'s body for one of [PHONE_ACTIONS], or null. */
    fun actBody(action: String): String? = when (action) {
        "extend" -> "{\"do\":\"extend\",\"minutes\":$EXTEND_MINUTES}"
        in PHONE_ACTIONS -> "{\"do\":\"$action\"}"
        else -> null
    }

    /** The buttons for a view, in order. */
    fun actionsOf(v: View): List<String> = when {
        !v.on -> emptyList()
        v.paused -> listOf("resume", "extend", "stop")
        else -> listOf("pause", "extend", "stop")
    }

    fun labelOf(action: String): String = when (action) {
        "pause" -> PAUSE
        "resume" -> RESUME
        "extend" -> EXTEND
        "stop" -> STOP
        else -> action
    }

    /** What the PC answered a start or an act, kept whole. */
    data class Reply(val code: Int, val body: JsonObject?)

    /** Whether it changed, and the PC's sentence to show. */
    fun said(reply: Reply): Pair<Boolean, String> {
        val b = reply.body
        return when {
            reply.code in 200..299 && b?.flag("ok") != false ->
                true to (b?.text("said") ?: "Done.")
            reply.code == 404 || reply.code == 501 -> false to MISSING
            else -> false to (b?.text("error") ?: "Not changed (HTTP ${reply.code}).")
        }
    }

    /** Whether a `focus` event should make the plate read again (every one does). */
    fun isFocusEvent(kind: String): Boolean = kind == "focus"

    private fun JsonObject.num(key: String): Double? =
        (this[key] as? JsonPrimitive)?.takeIf { it !is JsonNull && !it.isString }?.doubleOrNull
            ?.takeIf { it.isFinite() }

    private fun JsonObject.text(key: String): String? =
        (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull?.trim()?.takeIf { it.isNotEmpty() }

    private fun JsonObject.flag(key: String): Boolean? =
        (this[key] as? JsonPrimitive)?.takeIf { it !is JsonNull && !it.isString }?.booleanOrNull
}

File: jarvis-desktop/src/focus.js
/**
 * Focus sessions (the owner's decision of 2026-09-25; JARVIS-API.md section
 * 26; backend jarvis_focus.py).
 *
 * A timer plus Quiet. On this PC Jarvis watches which app or site is in
 * front, names a drift out loud, keeps only counts, and ends with a report
 * card. This module holds the words and reads what the PC sends; brain.js
 * draws the Work tab's "Focus session" (start, the countdown, Pause/Resume,
 * +10 minutes, Stop, the report card) and widget.js the small countdown on
 * the desktop that tints on a drift (Pause/Resume, Lock on, Stop). Both call
 * the Rust commands in src-tauri/src/brain/focus.rs.
 *
 * The PC never sends what was in front - only booleans and counts - and the
 * spoken line reaches this app as sound, fetched by Rust from the PC itself
 * (focus.rs play_callout), never as words.
 *
 * The phone says the same words (net/Focus.kt); tests/focus.mjs and the
 * phone's FocusTest check them against tools/gen_focus_cases.py's file.
 *
 * @module focus
 */

export const FOCUS_TITLE = "Focus session";
export const FOCUS_DETAIL =
  "A timer plus Quiet. On the PC, Jarvis watches which app or site is in front " +
  "and says so when you drift - out loud, on the PC only. It keeps counts, never " +
  "what it saw, and nothing leaves the PC. Off unless you start it.";
export const FOCUS_MISSING =
  "Your PC's Jarvis does not have focus sessions yet - run apply-patches.ps1 on the PC.";

export const DEFAULT_MINUTES = 25;
export const MIN_MINUTES = 1;
export const MAX_MINUTES = 240;
export const EXTEND_MINUTES = 10;
export const BAD_MINUTES = "A focus session is 1 to 240 minutes long.";

export const LABELS = {
  pause: "Pause",
  resume: "Resume",
  extend: "+10 minutes",
  stop: "Stop",
  lock: "Lock on",
};
export const START_LABEL = "Start";
export const LAST_TITLE = "Last session";
/** The widget's Lock on: what it does, because the button is IN Jarvis. */
export const LOCK_TITLE =
  "Go back to what you are working on - Jarvis locks on where you land. (Or say \"lock on this\" there.)";

/** Held on a stale link: they make Jarvis watch, or watch for longer. */
export const HELD_WHEN_STALE = new Set(["start", "resume", "extend", "lock"]);

const num = (v, d = 0) => (typeof v === "number" && Number.isFinite(v) ? v : d);
const text = (v) => (typeof v === "string" ? v.trim() : "");

/**
 * GET /api/focus as a view, or `{available: false, why}` for a PC without
 * focus sessions (Rust answers that for a 404 or 501) or anything unreadable.
 */
export function readFocus(body) {
  if (!body || typeof body !== "object" || body.available === false || typeof body.on !== "boolean") {
    return { available: false, why: text(body && body.why) || FOCUS_MISSING };
  }
  const r = body.report && typeof body.report === "object" ? body.report : null;
  return {
    available: true,
    on: body.on,
    paused: body.paused === true,
    state: text(body.state) || (body.on ? "locked" : "off"),
    minutes: num(body.minutes),
    left: Math.max(0, num(body.left_s)),
    intent: text(body.intent),
    deferred: body.deferred === true,
    onTarget: typeof body.on_target === "boolean" ? body.on_target : null,
    drifting: body.drifting === true,
    excused: body.excused === true,
    drifts: num(body.drifts),
    line: text(body.line),
    note: text(body.note),
    report: r ? {
      title: text(r.title),
      lines: Array.isArray(r.lines) ? r.lines.filter((x) => typeof x === "string") : [],
      clean: r.clean === true,
      streak: num(r.streak),
      completed: r.completed === true,
    } : null,
  };
}

/** Seconds left now, counted down from what the PC said `sinceMs` ago. */
export function leftNow(view, sinceMs = 0) {
  if (!view || !view.on) return 0;
  if (view.paused) return view.left;
  return Math.max(0, view.left - Math.floor(sinceMs / 1000));
}

/** "24:05", or "1:02:05" past an hour. */
export function clock(seconds) {
  const s = Math.max(0, Math.floor(seconds || 0));
  const h = Math.floor(s / 3600);
  const m = Math.floor((s % 3600) / 60);
  const sec = String(s % 60).padStart(2, "0");
  return h > 0 ? `${h}:${String(m).padStart(2, "0")}:${sec}` : `${m}:${sec}`;
}

/**
 * How the countdown looks: "drift" (tinted - off target, not excused),
 * "paused", "settling" (waiting to lock on), or "on".
 */
export function toneOf(view) {
  if (!view || !view.on) return "off";
  if (view.paused) return "paused";
  if (view.drifting && !view.excused) return "drift";
  if (view.deferred) return "settling";
  return "on";
}

/** The buttons, in order. `where`: "brain" or "widget" (the widget has Lock on). */
export function actionsOf(view, where = "brain") {
  if (!view || !view.on) return [];
  const first = view.paused ? "resume" : "pause";
  return where === "widget" ? [first, "lock", "stop"] : [first, "extend", "stop"];
}

/** The minutes typed, or null when they are not 1 to 240. */
export function minutesOf(value) {
  const s = String(value == null ? "" : value).trim();
  if (!/^\d{1,3}$/.test(s)) return null;
  const n = Number(s);
  return n >= MIN_MINUTES && n <= MAX_MINUTES ? n : null;
}

/** The drift count, in words. */
export function driftWords(n) {
  if (!n) return "No drifts so far.";
  return n === 1 ? "One drift so far." : `${n} drifts so far.`;
}
